mobile/src/test/java/org/openhab/habdroid/model/OpenHABWidgetTest.java

```diff
@@ -0,0 +1,418 @@
+    }
+
+    @Test
+    public void testGetService() throws Exception {
+        assertEquals("D", sut1.get(0).service());
+        assertEquals("XYZ", sut2.get(0).service());
+    }
+
+    @Test
+    public void testGetId() throws Exception {
+        assertEquals("demo", sut1.get(0).id());
+        assertEquals("demo", sut2.get(0).id());
+    }
+
+    @Test
+    public void testGetEncoding() throws Exception {
+        assertEquals("mpeg", sut1.get(0).encoding());
+        assertEquals(null, sut2.get(0).encoding());
+    }
+
+    private Node createXmlNode() throws Exception {
+        String xml = "" +
+                "<widget>" +
+                "  <widgetId>demo</widgetId>" +
+                "  <type>Group</type>" +
+                "  <label>Group1</label>" +
+                "  <icon>groupicon</icon>" +
+                "  <url>http://localhost/url</url>" +
+                "  <minValue>0.0</minValue>" +
+                "  <maxValue>10.0</maxValue>" +
+                "  <step>1</step>" +
+                "  <refresh>10</refresh>" +
+                "  <period>D</period>" +
+                "  <service>D</service>" +
+                "  <height>10</height>" +
+                "  <iconcolor>white</iconcolor>" +
+                "  <labelcolor>white</labelcolor>" +
+                "  <valuecolor>white</valuecolor>" +
+                "  <encoding></encoding>" +
+                "  <mapping>" +
+                "    <command>ON</command>\n" +
+                "    <label>On</label>" +
+                "  </mapping>" +
+                "  <item>" +
+                "    <type>GroupItem</type>" +
+                "    <name>group1</name>" +
+                "    <state>Undefined</state>" +
+                "    <link>http://localhost/rest/items/group1</link>" +
+                "  </item>" +
+                "  <linkedPage>" +
+                "    <id>0001</id>" +
+                "    <title>LinkedPage</title>" +
+                "    <icon>linkedpageicon</icon>" +
+                "    <link>http://localhost/rest/sitemaps/demo/0001</link>" +
+                "    <leaf>false</leaf>" +
+                "  </linkedPage>" +
+                "  <widget>" +
+                "    <widgetId>demo11</widgetId>" +
+                "    <type>Switch</type>" +
+                "  </widget>" +
+                "</widget>";
+        DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+        DocumentBuilder builder = dbf.newDocumentBuilder();
+        Document document = builder.parse(new InputSource(new StringReader(xml)));
+        Node rootNode = document.getFirstChild();
+        return rootNode;
+    }
+
+    /**
+     * @param id get different json objects depending on the id
+     *           1: All values are set
+     *           2: Different colors, no periode, max step size < 1, chart refresh < 100
+     *           3: Frame with Slider, Rollershutter switch and Rollershutter group
```

Comment:
Can you add and test
* setpoint with minValue > maxValue
* setpoint with stepSize < 0 ?

Reply:
Done.